src/Symfony/Component/Messenger/Transport/AmqpExt/Connection.php

```diff
@@ -94,13 +94,13 @@ public static function fromDsn(string $dsn, array $options = array(), bool $debu
     /**
      * @throws \AMQPException
      */
-    public function publish(string $body, array $headers = array()): void
+    public function publish(string $body, array $headers = array(), string $routingKey = null): void
     {
         if ($this->debug && $this->shouldSetup()) {
             $this->setup();
         }
 
-        $this->exchange()->publish($body, null, AMQP_NOPARAM, array('headers' => $headers));
+        $this->exchange()->publish($body, $routingKey, AMQP_NOPARAM, array('headers' => $headers));
```

Comment:
I'd suggest enhancing the DSN that creates the connection object so that one could express/decide to ignore the routing key at "send time" and force routing without any routing key.
The purpose would be to force messages to go to a workers without taking care of the routing key - and *then* the worker would use the stamp to decide for the next hop destination.

Reply:
I'm not sure to understand. You prefer to leave the `routing_key` parameter on the publish method to null ?

Your idea is to have multiple transports configured ? For example if we want to handle log messages :

```yml
framework:
    messenger:
        transports:
            log.info: 
                dsn: '%env(MESSENGER_TRANSPORT_DSN)%?routing_key=log.info'
                options:
                    exchange: {name: "log", type: "topic"}
                    queue: {name: "log.info", routing_key: "log.info"}
            log.error: 
                dsn: '%env(MESSENGER_TRANSPORT_DSN)%?routing_key=log.error'
                options:
                    exchange: {name: "log", type: "topic"}
                    queue: {name: "log.error", routing_key: "log.error"}

        routing:
            # Route your messages to the transports
            'App\Message\LogMessage': [log.info, log.error]
```

A message would be sent to multiple transport and then each transport will decide if he need to handle the message or not ?

RabbitMQ already do this for us by routing the received message using the routing_key. We can then have only one transport for sending and one transport per queue for receiving.

```yml
framework:
    messenger:
        transports:
            # to send any log message
            log:
                dsn: '%env(MESSENGER_TRANSPORT_DSN)%'
                options:
                    exchange: {name: "log", type: "topic"}
                    queue: {name: "log.info", routing_key: "log.*"}
            # Only for receiving using the messenger:consume-messages command
            log.info: 
                dsn: '%env(MESSENGER_TRANSPORT_DSN)%'
                options:
                    exchange: {name: "log", type: "topic"}
                    queue: {name: "log.info", routing_key: "log.info"}
            log.error: 
                dsn: '%env(MESSENGER_TRANSPORT_DSN)%'
                options:
                    exchange: {name: "log", type: "topic"}
                    queue: {name: "log.error", routing_key: "log.error"}

        routing:
            # Route your messages to the transports
            'App\Message\LogMessage': [log]
```

And the developer just need to create the message with its routing_key

```php
// The routing key is computed regarding the severity of the log message
$bus->dispatch(new Envelop(new \App\Message\LogMessage(...), new RoutingKeyStamp('log.error')));
```

I understand that it is thightly coupled with AMQP and therefor this is not agnostic :thinking: